packages/aws_signature_v4/lib/src/signer/aws_signer.dart

```diff
@@ -0,0 +1,308 @@
+      algorithm: algorithm,
+      credentialScope: credentialScope,
+      canonicalRequest: canonicalRequest,
+    );
+    final seedSignature = algorithm.sign(sts, signingKey);
+    final signedBody = serviceConfiguration.signBody(
+      algorithm: algorithm,
+      contentLength: contentLength,
+      signingKey: signingKey,
+      seedSignature: seedSignature,
+      credentialScope: credentialScope,
+      canonicalRequest: canonicalRequest,
+    );
+
+    return _buildSignedRequest(
+      credentials: credentials,
+      credentialScope: credentialScope,
+      signature: seedSignature,
+      body: signedBody,
+      contentLength: contentLength,
+      canonicalRequest: canonicalRequest,
+    );
+  }
+
+  /// Creates the string-to-sign (STS) for the canonical request.
+  @visibleForTesting
+  String stringToSign({
+    required AWSAlgorithm algorithm,
+    required AWSCredentialScope credentialScope,
+    required CanonicalRequest canonicalRequest,
+  }) {
+    final sb = StringBuffer();
+    sb.writeln(algorithm);
+    sb.writeln(credentialScope.dateTime);
+    sb.writeln(credentialScope);
+    sb.write(canonicalRequest.hash);
+
+    return sb.toString();
+  }
+
+  /// Creates an authorization header for a signed request.
+  @visibleForTesting
+  String createAuthorizationHeader({
+    required AWSCredentials credentials,
+    required AWSCredentialScope credentialScope,
+    required SignedHeaders signedHeaders,
+    required String signature,
+  }) {
+    return [
+      algorithm.id,
+      'Credential=${credentials.accessKeyId}/$credentialScope,',
+      'SignedHeaders=$signedHeaders,',
+      'Signature=$signature',
+    ].join(' ');
+  }
+
+  /// Builds a signed request from [canonicalRequest] and [signature].
+  AWSSignedRequest _buildSignedRequest({
+    required AWSCredentials credentials,
+    required CanonicalRequest canonicalRequest,
+    required String signature,
+    required Stream<List<int>> body,
+    required int contentLength,
+    required AWSCredentialScope credentialScope,
+  }) {
+    // The signing process requires component keys be encoded. However, the
+    // actual HTTP request should have the pre-encoded keys.
+    final queryParameters = Map.of(canonicalRequest.queryParameters);
+
+    // Similar to query parameters, some header values are canonicalized for
+    // signing. However their original values should be included in the
+    // headers map of the HTTP request.
+    final headers = Map.of(canonicalRequest.headers);
+
+    // If the session token was omitted from signing, include it now.
+    final sessionToken = credentials.sessionToken;
+    final includeSessionToken =
+        sessionToken != null && canonicalRequest.omitSessionTokenFromSigning;
+    if (canonicalRequest.presignedUrl) {
+      queryParameters[AWSHeaders.signature] = signature;
+      // TODO(dnys1): Remove second check when package SDK dep is >=2.15.0
+      if (includeSessionToken && sessionToken != null) {
```

Comment:
Does the "second check" refer to the `sessionToken != null` part of the condition?

Reply:
Yes

Reply:
I see. I'm still confused why that means it needs to be here when `includeSessionToken` being `true` should already mean that `sessionToken` is not null. 😓

Reply:
There's an analyzer change in 2.15 which picks up that sessionToken should not be null. In <2.15, the check is required. Here's what it looks like in 2.15:
<img width="629" alt="Screen Shot 2022-04-20 at 11 04 40 AM" src="https://user-images.githubusercontent.com/24740863/164294480-aa9e2b53-2c1e-4ae1-ae11-6d802d64d5f4.png">